scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala

```diff
@@ -769,18 +769,21 @@ sealed trait SCollection[T] extends PCollectionWrapper[T] {
     out: () => PrintStream = () => Console.out,
     prefix: String = "",
     enabled: Boolean = true
-  ): SCollection[T] =
+  )(implicit coder: Coder[T]): SCollection[T] =
     if (enabled) {
-      this.filter { e =>
-        out().println(prefix + e)
-
-        // filter that never removes
-        true
-      }
+      tap(elem => out().println(prefix + elem))
     } else {
       this
     }
 
+  /**
+   * Applies f to each element of this [[SCollection]], and returns the original value.
+   *
+   * @group debug
+   */
+  def tap[U](f: T => U)(implicit coder: Coder[T]): SCollection[T] =
+    map { elem => f(elem); elem }
```

Comment:
alternately we could implement it like

```
  def tap[U](f: T => U)(implicit coder: Coder[T]): SCollection[T] = {
    map(f)
    this
}
```

so that `map(f)` would be run in parallel with the next transform a user specifies instead of having to complete `f(elem)` for entire collection before moving on to the next step? Since the return type of `f` is discarded it seems safe to assume it can be a leaf-node in the job graph? wdyt?

also does `U` need to be parameterized? Can't `f` just be `T => Any` or `T => Unit` ?

Reply:
so `tap` is actually kinda of particular since it allows `side-effects` to take place. As such you might or might not require `f` to complete before passing to the next step. I think i'm leaning towards letting the user make that decision. ¯\_(ツ)_/¯ 

by having a type param i'm basically saying almost the same thing as `T => Any` but I'm not swallowing the type, it's nicer overall (better inference). `T => Unit` is to restrictive as it would force users to actually pass a function of that type if they are not creating an in-place lambda.